### PR TITLE
binlog proto: add proto file

### PR DESCRIPTION
**Pull request by july2993**
part of https://github.com/pingcap/tidb-tools/pull/21, contains only files about protobuf

### CODE REVIEW

**Comment by july2993** [comment]
@GregoryIan PTAL

File: obinlog/proto/binlog.proto [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+// for text and char type, string is set
+// for blob and binary type, bytes is set
+// for enum, set, uint64 is set
+// for json, bytes is set
+message Column {
+	optional bool is_null = 1 [default = false];
+
+    optional int64 int64_value = 2;
+	optional uint64 uint64_value = 3;
```
Comment (IANTHEREAL) [inline obinlog/proto/binlog.proto]: format code

File: obinlog/proto/binlog.proto [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	repeated Row changed_rows = 6;
+}
+
+message DMLData {
+    // tables contains all the table changes.
+    repeated Table tables      = 1;
+}
+message DDLData {
+	// the current database use
+	optional string schema_name = 1;
```
Comment (IANTHEREAL) [inline obinlog/proto/binlog.proto]: how about adding table_name, if user want to route table.

File: obinlog/proto/binlog.proto [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+message Row {
+	repeated Column columns = 1;
+}
+
+enum MutationType {
+	Insert = 0;
+	Update = 1;
+	Delete = 4;
```
Comment (IANTHEREAL) [inline obinlog/proto/binlog.proto]: 2

File: obinlog/proto/binlog.proto [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    optional double double_value = 4;
+    optional bytes bytes_value = 6;
+    optional string string_value = 7;
+}
+
+message ColumnInfo {
+    optional string  name       = 1 [(gogoproto.nullable) = false];
+    // column field type in mysql
+    optional string  mysql_type = 3 [(gogoproto.nullable) = false];
+	optional bool is_primary_key = 4 [(gogoproto.nullable) = false];
```
Comment (IANTHEREAL) [inline obinlog/proto/binlog.proto]: is_unique and is_primary?

Comment (IANTHEREAL) [reply obinlog/proto/binlog.proto]: the reason I say that is I think we try to restore it as much as possible

Comment (july2993) [reply obinlog/proto/binlog.proto]: it's is_primary,  maybe we can add more info later

**Comment by iamxy** [comment]
Two questions first:
1. What does the `obinlog` mean?
2. What's the purpose of this PR?

Please follow the [Guidelines](https://github.com/pingcap/tidb/blob/master/pull_request_template.md) to describe the details about that.

**Comment by july2993** [comment]
> Two questions first:
>
> What does the obinlog mean?
> What's the purpose of this PR?
> Please follow the Guidelines to describe the details about that.

 just part of https://github.com/pingcap/tidb-tools/pull/21 for GregoryIan's request to easy review, add some describle in the relate pr, please see the relate pr

**Comment by IANTHEREAL** [comment]
according to the discussion, the name of the binlog related module are:
* `binlog-service` =>  service that support binlog subscriptions
* `binlog-driver` => library that support binlog protocol parsing and reading (from kafka)
* `binlog-proto` => format protocol of binlog event

File: obinlog/proto/binlog.proto (hunk 2 above)
Comment (IANTHEREAL) [reply obinlog/proto/binlog.proto]: add it

File: obinlog/proto/binlog.proto [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+// for blob and binary type, bytes is set
+// for enum, set, uint64 is set
+// for json, bytes is set
+message Column {
+    optional bool is_null = 1 [default = false];
+
+    optional int64 int64_value = 2;
+    optional uint64 uint64_value = 3;
+    optional double double_value = 4;
+    optional bytes bytes_value = 6;
```
Comment (IANTHEREAL) [inline obinlog/proto/binlog.proto]: 5? check and fix similar problems

Comment (july2993) [reply obinlog/proto/binlog.proto]: i think it's ok with this,  no need to use continuous number and should not change, but i can change it now as no servise use this proto now

Comment (IANTHEREAL) [reply obinlog/proto/binlog.proto]: I know it's ok,  but ugly

**Comment by july2993** [comment]
@WangXiangUSTC PATL and the relate PR

File: binlog_proto/proto/binlog.proto [hunk 6]
```diff
@@ -0,0 +1,83 @@
+syntax = "proto2";
+package binlog;
+import "gogoproto/gogo.proto";
+
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.sizer_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+
+// for date and time type, string is set
```
Comment (WangXiangUSTC) [inline binlog_proto/proto/binlog.proto]: how about change to "for date and time type,  set `string_value`"?

File: binlog_proto/proto/binlog.proto [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+// for int type int64 or uint64 is set
+// for float, double, double is set
+// for decimal, string is set in human-readable now
+// for bit, bytes is set
+
+// for text and char type, string is set
+// for blob and binary type, bytes is set
+// for enum, set, uint64 is set
+// for json, bytes is set
+message Column {
```
Comment (WangXiangUSTC) [inline binlog_proto/proto/binlog.proto]: this struct name seems not suitable, how about change a name

File: binlog_proto/proto/binlog.proto [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    optional uint64 uint64_value = 3;
+    optional double double_value = 4;
+    optional bytes bytes_value = 5;
+    optional string string_value = 6;
+}
+
+message ColumnInfo {
+    optional string  name       = 1 [(gogoproto.nullable) = false];
+    // column field type in mysql
+    optional string  mysql_type = 2 [(gogoproto.nullable) = false];
```
Comment (WangXiangUSTC) [inline binlog_proto/proto/binlog.proto]: how about add a enum to list the mysql type

File: binlog_proto/proto/binlog.proto [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+enum BinlogType {
+    DML = 0; // has commit_ts, dml_data
+    DDL = 1; // has commit_ts, ddl_query
+}
+
+// Binlog contains all the changes in a transaction.
+message Binlog {
+    optional BinlogType    type           = 1 [(gogoproto.nullable) = false];
+    optional int64         commit_ts      = 2 [(gogoproto.nullable) = false];
+    // dml_data is marshalled from DML type,
```
Comment (WangXiangUSTC) [inline binlog_proto/proto/binlog.proto]: ',' -> '.'

File: binlog_proto/proto/binlog.proto [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+}
+message DDLData {
+    // the current database use
+    optional string schema_name = 1;
+    // ddl_query is the original ddl statement query.
+    optional bytes  ddl_query  = 2;
+}
+
+enum BinlogType {
+    DML = 0; // has commit_ts, dml_data
```
Comment (WangXiangUSTC) [inline binlog_proto/proto/binlog.proto]: what is the mean of "has commit_ts"

File: binlog_proto/proto/binlog.proto (hunk 7 above)
Comment (july2993) [reply binlog_proto/proto/binlog.proto]: field? i think column is ok

File: binlog_proto/proto/binlog.proto (hunk 8 above)
Comment (july2993) [reply binlog_proto/proto/binlog.proto]: i think we can just use the string same with mysql, but i will improve the comment

File: binlog_proto/proto/binlog.proto (hunk 10 above)
Comment (july2993) [reply binlog_proto/proto/binlog.proto]: commit_ts field in the Binlog message, but seems no need to specify it, i will remove it

**Comment by july2993** [comment]
@WangXiangUSTC  PTAL

**Comment by WangXiangUSTC** [comment]
please format the code and comment, REST LGTM

**Comment by july2993** [comment]
@GregoryIan PTAL

**Comment by IANTHEREAL** [comment]
LGTM

**Pull request merged**